drivers/include/periph/hwcrypto.h

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (C) 2016 Bas Stottelaar <basstottelaar@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_periph_hwcrypt Hardware Crypto
+ * @ingroup     drivers_periph
+ * @brief       Low-level hardware crypto peripheral driver
+ *
+ * Modern CPUs may be equipped with a hardware crypto peripheral to speed up
+ * cryptographical operations such as (as)symmetric ciphers and hashing
+ * algorithms. With no or little intervention, the CPU can typically execute
+ * more operations than the software alternative.
+ *
+ * This driver provides an abstract for the different ciphers and hashes that
+ * may be supported. Using HAVE_HWCRYPTO_xxx defines, software can choose to
+ * use the hardware crypto alternative, or fall back to a software variant.
+ *
+ * Current interface is optimized for (as)symmetric block ciphers and block
+ * hashes.
+ *
+ * All operations block until they are finished. Operations may change the
+ * internal peripheral state (e.g. load the key, keep the digest interally),
+ * therefore must be used to control mutual exclusive access.
+ *
+ * @{
+ * @file
+ * @brief       Low-level hardware crypto driver interface definitions
+ *
+ * @author      Bas Stottelaar <basstottelaar@gmail.com>
+ */
+
+#ifndef HWCRYPTO_H
+#define HWCRYPTO_H
+
+#include <stdint.h>
+
+#include "periph_cpu.h"
+#include "periph_conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* guard file in case no HWCRYPTO device is defined */
+#if HWCRYPTO_NUMOF
+
+/**
+ * @brief   Supported ciphers.
+ *
+ * Define HAVE_HWCRYPTO_xxx in your CPUs periph_cpu.h if needed.
+ *
+ * Extend at will.
+ */
+typedef enum {
+#ifdef HAVE_HWCRYPTO_AES128
+    HWCRYPTO_AES128,                  /**< AES128 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_AES256
+    HWCRYPTO_AES256,                  /**< AES256 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_DES
+    HWCRYPTO_DES,                     /**< DES cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_3DES
+    HWCRYPTO_3DES,                    /**< 3DES cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_TWOFISH
+    HWCRYPTO_TWOFISH,                 /**< Two-fish cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA512
+    HWCRYPTO_RSA512,                  /**< RSA512 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA1024
+    HWCRYPTO_RSA1024,                 /**< RSA1024 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA2048
+    HWCRYPTO_RSA2048,                 /**< RSA2048 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC128
+    HWCRYPTO_ECC128,                  /**< ECC128 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC192
+    HWCRYPTO_ECC192,                  /**< ECC192 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC256
+    HWCRYPTO_ECC256,                  /**< ECC256 cipher */
+#endif
+    HWCRYPT_CIPHER_NULL,
```

Comment:
Wouldn't it be semantically nicer if HWCRYPT_x_NULL was the first entry, so NULL == 0?